### PR TITLE
Add test for parquet loader.

### LINKED ISSUE #760: Test for parquet loader because type system has changed
We need to verify parquet loader works with the .NET standard type system.

### PR DESCRIPTION
**Pull request by codemzs**
fixes #760

### CODE REVIEW

File: test/Microsoft.ML.TestFramework/TestInitialization.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using Xunit;
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/TestInitialization.cs]: Is this change necessary?

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs [hunk 2]
```diff
@@ -145,4 +145,24 @@ public void TestLdaTransformEmptyDocumentException()
             Assert.True(false, "The LDA transform does not throw expected error on empty documents.");
         }
     }
+
+    public sealed partial class TestDataPipe : TestDataPipeBase
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs]: IMO, we should have 1 test class per file.

File: test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs [hunk 3]
```diff
@@ -145,4 +145,24 @@ public void TestLdaTransformEmptyDocumentException()
...
+    {
+
+        [Fact]
+        public void TestParquetPrimitiveDataTypes()
+        {
+            string pathData = GetDataPath(@"..\data\Parquet", "alltypes.parquet");
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/DataPipe/TestDataPipe.cs]: `GetDataPath` is already at the `test\data` folder.  You shouldn't need `..\data\` since that just gets you back into the same folder.  You should just have `GetDataPath("Parquet", "alltypes.parquet");`

File: test/BaselineOutput/SingleRelease/SavePipe/TestParquetNull-Data.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#@ TextLoader{
```
Comment (eerhardt) [inline test/BaselineOutput/SingleRelease/SavePipe/TestParquetNull-Data.txt]: I wish we had a mechanism in place so we didn't have to duplicate baseline files between Debug and Release when the files were identical.

Comment (codemzs) [reply test/BaselineOutput/SingleRelease/SavePipe/TestParquetNull-Data.txt]: Me too but unfortunately beyond the scope of this change. Lets open an issue if there isn't one open. 

---
In reply to: [213393285](https://github.com/dotnet/machinelearning/pull/757#discussion_r213393285) [](ancestors = 213393285)

Comment (eerhardt) [reply test/BaselineOutput/SingleRelease/SavePipe/TestParquetNull-Data.txt]: Can you open one?

Comment (codemzs) [reply test/BaselineOutput/SingleRelease/SavePipe/TestParquetNull-Data.txt]: Done. #758

**Comment by shauheen** [comment]
Can you please associate this to an issue?

File: test/BaselineOutput/SingleRelease/SavePipe/TestParquetPrimitiveDataTypes-Data.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+#@   col=Amount2:R8:8
+#@   col=Flag:U1:9
+#@   col=Flag2:I1:10
+#@   col=Flag3:I2:11
+#@   col=Flag4:U2:12
+#@   col=Flag5:TS:13
+#@ }
+Id	Timestamp	Message	Data	IsDeleted	Amount	TotalAmount	Counter	Amount2	Flag	Flag2	Flag3	Flag4	Flag5
+1	"2000-01-01T01:01:01.0000000+00:00"	Record1	SomeData3	0	125.4	400	300000	3331313	3	-3	-600	600	"3100.00:00:00.1000000"
+1	"2000-12-31T23:59:59.9990000+00:00"	Record2	SomeData4	0	126.4	500	400000	4331313	4	-4	-700	700
```
Comment (TomFinley) [inline test/BaselineOutput/SingleRelease/SavePipe/TestParquetPrimitiveDataTypes-Data.txt]: >	 [](start = 96, length = 1)

The thing following this appears to be a missing value, from the fact that the field is blank. There must be something wrong *somewhere* it seems to me, since if it is not nullable, why is it mapping to the missing value?

Also there is the point that in your second datetime PR, if it *is* nullable, then this test will throw, and we ought to have some "happy path" test where we can just baseline the outputs of converting Parquet to test (e.g., `savedata data=parquet-file.parquet dout=output.txt`), which we cannot do if we are, by design, throwing exceptions. Both of these files at least at first glance seem to be converting to text files where there are missing values, either because there are missing values there, or else some fault in the parquet loader.

Comment (codemzs) [reply test/BaselineOutput/SingleRelease/SavePipe/TestParquetPrimitiveDataTypes-Data.txt]: Hi @TomFinley , It maps to missing when there is an overflow. Please see the below conversion code. Here the value isn't nullable. In DvDateTime code I have changed this function to set dst to default because we decided to not throw on reading missing values anywhere but rather set minimum or default and we also decided we could switch between default and minimum from type to type (i.e for int we use minimum and for boolean we use default) as long as we are consistent with throwing exception vs not throwing exception. 

            /// <summary>
            /// Converts a Parquet Interval data type value to a DvTimeSpan data type value.
            /// </summary>
            /// <param name="src">Parquet Interval value (int : months, int : days, int : milliseconds).</param>
            /// <param name="dst">DvTimeSpan object.</param>
            public void Conv(ref Interval src, ref DvTimeSpan dst)
            {
                try
                {
                    dst = new DvTimeSpan(TimeSpan.FromDays(src.Months * 30 + src.Days) + TimeSpan.FromMilliseconds(src.Millis));
                }
                catch (Exception ex)
                {
                    // Handle TimeSpan OverflowException
                    _ch.Error("Cannot convert Inteval to DvTimeSpan. Exception : '{0}'", ex.Message);
                    dst = DvTimeSpan.NA;
                }
            }

Comment (TomFinley) [reply test/BaselineOutput/SingleRelease/SavePipe/TestParquetPrimitiveDataTypes-Data.txt]: Very well. As you said in an offline discussion this ought to be a configurable option, but possibly not as part of this PR, since we will have a chance to revisit post 0.5.

**Comment by TomFinley** [comment]
> 

We still need to resolve this, can we fix it so we can get this in?

---
Refers to: test/BaselineOutput/SingleDebug/SavePipe/TestParquetPrimitiveDataTypes-Data.txt:22 in 96ea4cc. [](commit_id = 96ea4cc3bbf9820775a3e128e7100fd33bd2e18a, deletion_comment = False)
